src/actix/api/snapshot_api.rs

```diff
@@ -333,24 +333,34 @@ async fn upload_shard_snapshot(
     query: web::Query<SnapshotUploadingParam>,
     MultipartForm(form): MultipartForm<SnapshottingForm>,
 ) -> impl Responder {
+    let (collection, shard) = path.into_inner();
     let SnapshotUploadingParam { wait, priority } = query.into_inner();
 
-    let future = async move {
-        let (collection, shard) = path.into_inner();
-        let collection = toc.get_collection(&collection).await?;
-        collection.assert_shard_exists(shard).await?;
+    let future = cancel::future::on_drop(move |cancel| async move {
+        let future = async {
+            let collection = toc.get_collection(&collection).await?;
+            collection.assert_shard_exists(shard).await?;
+
+            Result::<_, helpers::HttpError>::Ok(collection)
+        };
+
+        let collection = cancel::future::on_token(cancel.clone(), future).await??;
 
+        // `recover_shard_snapshot_impl` is *not* cancel-safe!
         common::snapshots::recover_shard_snapshot_impl(
             &toc,
             &collection,
             shard,
             form.snapshot.file.path(),
             priority.unwrap_or_default(),
+            cancel,
```

Comment:
Let's make cancellation optional for the REST and user-facing API. So that timing out of the recovery/upload requests are still result in complete recovery in the background by default. (this should preserve backward compatibility)

The internal API might be cancellable by default

Reply:
Done.